docker/Dockerfile.cpp

```diff
@@ -0,0 +1,91 @@
+#
+# NOTE: To build this you will need a docker version > 18.06 with
+#       experimental enabled and DOCKER_BUILDKIT=1
+#
+#       If you do not use buildkit you are not going to have a good time
+#
+#       For reference:
+#           https://docs.docker.com/develop/develop-images/build_enhancements/
+
+
+ARG BASE_IMAGE=ubuntu:20.04
+ARG PYTHON_VERSION=3.9
+ARG CMAKE_VERSION=3.26.4
+ARG BRANCH_NAME="master"
+ARG USE_CUDA_VERSION=""
+
+FROM ${BASE_IMAGE} AS cpp-dev-image
+ARG BASE_IMAGE
+ARG PYTHON_VERSION
+ARG CMAKE_VERSION
+ARG BRANCH_NAME
+ARG USE_CUDA_VERSION
+ENV PYTHONUNBUFFERED TRUE
+
+RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt \
+    apt-get update && \
+    apt-get install software-properties-common -y && \
+    add-apt-repository -y ppa:deadsnakes/ppa && \
+    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
+        sudo \
+        vim \
+        git \
+        curl \
+        wget \
+        rsync \
+        gpg \
+        ca-certificates \
+        lsb-release \
+        openjdk-17-jdk \
+        python$PYTHON_VERSION \
+        python$PYTHON_VERSION-dev \
+        python$PYTHON_VERSION-venv \
+    && rm -rf /var/lib/apt/lists/*
+
+# Create a virtual environment and "activate" it by adding it first to the path.
+RUN python$PYTHON_VERSION -m venv /home/venv
+ENV PATH="/home/venv/bin:$PATH"
+
+# Enable installation of recent cmake release
+# Ref: https://apt.kitware.com/
+RUN (wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | gpg --dearmor - | tee /usr/share/keyrings/kitware-archive-keyring.gpg >/dev/null) \
+    && (echo "deb [signed-by=/usr/share/keyrings/kitware-archive-keyring.gpg] https://apt.kitware.com/ubuntu/ $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/kitware.list >/dev/null) \
+    && apt-get update \
+    && (test -f /usr/share/doc/kitware-archive-keyring/copyright || sudo rm /usr/share/keyrings/kitware-archive-keyring.gpg) \
+    && sudo apt-get install kitware-archive-keyring \
+    && rm -rf /var/lib/apt/lists/*
+
+# Pin cmake and cmake-data version
+# Ref: https://manpages.ubuntu.com/manpages/xenial/man5/apt_preferences.5.html
+RUN echo "Package: cmake\nPin: version $CMAKE_VERSION*\nPin-Priority: 1001" > /etc/apt/preferences.d/cmake
+RUN echo "Package: cmake-data\nPin: version $CMAKE_VERSION*\nPin-Priority: 1001" > /etc/apt/preferences.d/cmake-data
+
+# Install CUDA toolkit to enable "libtorch" build with GPU support
+RUN apt-get update && \
+    if echo "$BASE_IMAGE" | grep -q "cuda:"; then \
+        if [ "$USE_CUDA_VERSION" = "cu121" ]; then \
+            apt-get -y install cuda-toolkit-12-1; \
+        elif [ "$USE_CUDA_VERSION" = "cu118" ]; then \
+            apt-get -y install cuda-toolkit-11-8; \
+        else \
+            echo "Cuda version not supported by CPP backend: $USE_CUDA_VERSION"; \
+            exit 1; \
+        fi; \
+    fi \
+    && rm -rf /var/lib/apt/lists/*
```

Comment:
All of this part should be replaced with install_dependency (ref: https://github.com/pytorch/serve/blob/master/docker/Dockerfile.dev#L71)

Reply:
Additional packages aside from the ones installed by `install_depenedencies.py` are required. Following are the error traces I see, if the `cuda-toolkit` is not installed:

```
[2024-03-11 23:50:09,812] torch._inductor.utils: [WARNING] not enough SMs to use max_autotune_gemm mode
In file included from /home/venv/lib/python3.9/site-packages/torch/include/torch/csrc/inductor/aoti_runtime/model.h:15,
                 from /home/venv/lib/python3.9/site-packages/torch/include/torch/csrc/inductor/aoti_runtime/model_container.h:13,
                 from /serve/examples/cpp/aot_inductor/bert/cly46ndmfcer53dv4xkdyjmpl3mc7277slw3od3ue5ygudxcb766.cpp:2:
/home/venv/lib/python3.9/site-packages/torch/include/torch/csrc/inductor/aoti_runtime/device_utils.h:14:10: fatal error: cuda.h: No such file or directory
   14 | #include <cuda.h>
      |          ^~~~~~~~
compilation terminated.
```

Installing `cuda-compiler` resolves the above issue, but we'll then run into:
```
-- Found Threads: TRUE
CMake Error at /home/venv/lib/python3.9/site-packages/torch/share/cmake/Caffe2/public/cuda.cmake:70 (message):
  Failed to find nvToolsExt
Call Stack (most recent call first):
  /home/venv/lib/python3.9/site-packages/torch/share/cmake/Caffe2/Caffe2Config.cmake:87 (include)
  /home/venv/lib/python3.9/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:25 (find_package)
```

Instead of manually having to install all the necessary packages individually, it is cleaner to install the `cuda-toolkit`.